### PR TITLE
Improves test support of ASR domain, updation of vocabulary and various fixes

### PR DESCRIPTION
**Pull request by titu1994**
# Changelog
- Adds `test_ds` support for all experimental model configs
- Adds `trainer.test` support for CTC and CTC BPE models.
  - Attaches trainer to "ready" model that has been validated for `trainer.test()`
- Adds ASRModel level support for multi dataset test set evaluation
- Adds support for changing label set for ASR Classification models vis `change_labels(new_labels)`
- Refactor Tokenizer setup for BPE models
- Adds support for changing tokenizer for ASR BPE models via `change_vocabulary(new_tokenizer_dir, new_tokenizer_type)`
  - Signature mismatch with base class. Options are to 
  - 1) override base class method, raise not implemented and direct to use another method
  - 2) continue with the signature mismatch in place (not a pythonic solution).
- Adds tests for new functions

# Bugfixes
- Corrects an issue in ModelPT.py where self._validation_loss_idx is used for multi test set evaluation
- Replaces Value Error for scheduler setup - required during test step (without attaching any train dataset).
  - Schedulers will now log a warning instead
- `model_utils` will now do a two step check - once to check if _trainer is available, and if _trainer.fast_dev_run is set.
  - Avoids a crash where any one of the 4 cases is None or false or does not exist in the dictionary at all.

## Reason for warning instead of raising error in scheduler setup
Trainer.test() calls optimizer + scheduler setup (it calls the internal fit call with different parameters). Since the fit call has the configure_optimizers() call, the scheduler is attempted to be built.

If the model is tested right after training, then there's no issues as the train ds is in place and can be used to recreate the scheduler even if it's not used.

But if you restore and immediately use trainer.test(), then the train ds does not exist (even though the user has already setup test_ds). This will then raise the above error. In order to avoid raising an error in a valid use case - testing without setting up a training dataset from restoration, this needs to be a warning instead of an error.

### CODE REVIEW

File: nemo/core/optim/lr_scheduler.py [hunk 1]
```diff
@@ -496,14 +496,22 @@ def prepare_lr_scheduler(
...
         if train_dataloader is None:
-            raise ValueError(
-                'As `iters_per_batch` is provided, it is required to pass the train dataloader in order '
-                'to compute effective maximum number of steps'
+            logging.warning(
+                'As `iters_per_batch` is provided/computed, it is required to pass the train dataloader in order\n'
+                'to compute effective maximum number of steps.\n'
+                'Scheduler will not be instantiated !'
             )
+            return None
```
Comment (blisc) [inline nemo/core/optim/lr_scheduler.py:504]: What's the reasoning for this change?

File: nemo/collections/asr/models/ctc_bpe_models.py [hunk 2]
```diff
@@ -212,6 +216,56 @@ def _setup_transcribe_dataloader(self, config: Dict) -> 'torch.utils.data.DataLo
         temporary_datalayer = self._setup_dataloader_from_config(config=DictConfig(dl_config))
         return temporary_datalayer
 
+    def change_vocabulary(self, new_tokenizer_dir: str, new_tokenizer_type: str):
```
Comment (blisc) [inline nemo/collections/asr/models/ctc_bpe_models.py:219]: Can we introduce a CI test that tests `change_vocabulary()` works with fine-tuning?

**Review by percy21081993 — APPROVED** [review]
AVE

**Review by percy21081993 — COMMENTED** [review]
VNS001

File: nemo/core/optim/lr_scheduler.py (hunk 1 above)
Comment (titu1994) [reply nemo/core/optim/lr_scheduler.py:504]: Trainer.test() calls optimizer + scheduler setup (it calls the internal fit call with different parameters). Since the fit call has the configire_optimizers() call, the scheduler is attempted to be built. 

Of the model is tested right after training, then there's no issues as the train ds is in place and can be used to recreate the scheduler even if it's not used. 

But if you restore and immediately use trainer.test)), then the train ds does not exist (even though the user has already setup test_ds()). This will then raise the above error. In order to avoid raising an error in a valid use case - testing without setting up a training dataset from restoration, this needs to be a warning instead of an error.

File: nemo/collections/asr/models/ctc_bpe_models.py (hunk 2 above)
Comment (titu1994) [reply nemo/collections/asr/models/ctc_bpe_models.py:219]: As in a Jenkins test ? Hmm we'd need to write a new train script for that support. As in the standard speech_to_text_bpe.py won't be able to handle that.

Comment (titu1994) [reply nemo/collections/asr/models/ctc_bpe_models.py:219]: There is a unittest that replaces the vocabulary though, if that's what you meant.

Comment (blisc) [reply nemo/collections/asr/models/ctc_bpe_models.py:219]: I saw the unittest which is good, but I feel like we are missing a higher level test that makes sure everything works as intended during fine-tuning

File: nemo/core/optim/lr_scheduler.py (hunk 1 above)
Comment (blisc) [reply nemo/core/optim/lr_scheduler.py:504]: I though `configure_optimizers` was only called during trainer.fit(), is this wrong?

Comment (titu1994) [reply nemo/core/optim/lr_scheduler.py:504]: This is right, but PTL actually calls internally trainer.fit() even for trainer.test() (but with different flags preset so as to force eval mode execution)

File: nemo/collections/asr/models/ctc_bpe_models.py (hunk 2 above)
Comment (titu1994) [reply nemo/collections/asr/models/ctc_bpe_models.py:219]: Hmm valid point. But then we need scripts for fine-tuning for all ASR + ASR BPE + classification models. What I'm planning is once this PR is done with reviews, to update the speech command notebook to have a section on fine-tuning with change of vocabulary. But that won't be run by CI which is a problem. 

Let's discuss a API spec for fine-tuning with the team and then add it to the ci tests
